File: war-ticket-example/flash_sale.php
<?php

session_start();

$redis = require 'redis.php';

$sessionId = session_id();

$limit = 1000;
$ttl = 900;

$now = time();
$expireAt = $now + $ttl;

$lua = file_get_contents('flash_sale.lua');

$result = $redis->eval(
    $lua,
    [
        'flash:active',
        "flash:hold:$sessionId",

        $now,
        $expireAt,
        $limit,
        $sessionId,
        $ttl
    ],
    2
);

$status = $result[0];
$count = $result[1];

switch ($status) {

    case 0:
        echo "Flash sale penuh. Active: $count";
        break;

    case 1:
        echo "Berhasil masuk flash sale. Active: $count";
        break;

    case 2:
        echo "Kamu sudah punya slot.";
        break;
}

File: war-ticket-example/flash_sale.lua
-- KEYS[1] = sorted set active key
-- KEYS[2] = hold key

-- ARGV[1] = current timestamp
-- ARGV[2] = expire timestamp
-- ARGV[3] = limit
-- ARGV[4] = session id
-- ARGV[5] = ttl seconds

local activeKey = KEYS[1]
local holdKey = KEYS[2]

local now = tonumber(ARGV[1])
local expireAt = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local sessionId = ARGV[4]
local ttl = tonumber(ARGV[5])

-- cleanup expired user
redis.call('ZREMRANGEBYSCORE', activeKey, '-inf', now)

-- hitung active user
local current = redis.call('ZCARD', activeKey)

if current >= limit then
    return {
        0,
        current
    }
end

-- cek apakah user sudah punya hold
local exists = redis.call('EXISTS', holdKey)

if exists == 1 then
    return {
        2,
        current
    }
end

-- simpan hold
redis.call('SET', holdKey, 1, 'EX', ttl)

-- masukkan ke sorted set
redis.call('ZADD', activeKey, expireAt, sessionId)

return {
    1,
    current + 1
}